Neos.EventSourcedNeosAdjustments/Classes/EventSourcedRouting/Projection/DocumentUriPathProjector.php

```diff
@@ -0,0 +1,527 @@
+            ]);
+        } catch (DBALException $e) {
+            throw new \RuntimeException(sprintf('Failed to insert root content stream identifier of the root workspace "%s": %s', $event->getWorkspaceName(), $e->getMessage()), 1599646608, $e);
+        }
+    }
+
+    public function whenRootNodeAggregateWithNodeWasCreated(RootNodeAggregateWithNodeWasCreated $event): void
+    {
+        if (!$this->isLiveContentStream($event->getContentStreamIdentifier())) {
+            return;
+        }
+        foreach ($event->getCoveredDimensionSpacePoints() as $dimensionSpacePoint) {
+            $this->insertNode([
+                'uriPath' => '',
+                'nodeAggregateIdentifierPath' => $event->getNodeAggregateIdentifier(),
+                'dimensionSpacePointHash' => $dimensionSpacePoint->getHash(),
+                'nodeAggregateIdentifier' => $event->getNodeAggregateIdentifier(),
+            ]);
+        }
+    }
+
+    public function whenNodeAggregateWithNodeWasCreated(NodeAggregateWithNodeWasCreated $event): void
+    {
+        if (!$this->isLiveContentStream($event->getContentStreamIdentifier())) {
+            return;
+        }
+        if (!$this->isDocumentNodeType($event->getNodeTypeName())) {
+            return;
+        }
+
+        $propertyValues = $event->getInitialPropertyValues()->getPlainValues();
+        $uriPathSegment = $propertyValues['uriPathSegment'] ?? '';
+
+        $shortcutTarget = null;
+        if ($this->isShortcutNodeType($event->getNodeTypeName())) {
+            $shortcutTarget = [
+                'mode' => $propertyValues['targetMode'] ?? 'firstChildNode',
+                'target' => $propertyValues['target'] ?? null,
+            ];
+        }
+
+        foreach ($event->getCoveredDimensionSpacePoints() as $dimensionSpacePoint) {
+            $parentNode = $this->tryGetNode(fn () =>$this->documentUriPathFinder->getOneByIdAndDimensionSpacePointHash($event->getParentNodeAggregateIdentifier(), $dimensionSpacePoint->getHash()));
+            if ($parentNode === null) {
+                // this should not happen
+                continue;
+            }
+            /** @var DocumentNodeInfo|null $precedingNode */
+            $precedingNode = null;
+
+            if ($event->getSucceedingNodeAggregateIdentifier() === null) {
+                $precedingNode = $this->tryGetNode(fn () => $this->documentUriPathFinder->getLastChildNode($parentNode->getNodeAggregateIdentifier(), $dimensionSpacePoint->getHash()));
+                if ($precedingNode !== null) {
+                    $this->updateNode($precedingNode, ['succeedingNodeAggregateIdentifier' => $event->getNodeAggregateIdentifier()]);
+                }
+            } else {
+                $precedingNode = $this->tryGetNode(fn () => $this->documentUriPathFinder->getPrecedingNode($event->getSucceedingNodeAggregateIdentifier(), $parentNode->getNodeAggregateIdentifier(), $dimensionSpacePoint->getHash()));
```

Comment:
... should we name the variables ...NodeInfo; to not confuse this with traversableNode?

Reply:
I find `precedingNode` easier to comprehend than `precedingNodeInfo` tbh and I don't think that there's a high risk for confusion since this projector has nothing to do with the other node models